// smashgg.js/src/v4/js/lib/models/Standing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Entrant_1 = require("./Entrant");
var Standing = /** @class */ (function () {
    function Standing(id, placement, entrant) {
        this.id = id;
        this.placement = placement;
        this.entrant = entrant;
    }
    Standing.getDefaultOptions = function () {
        return {
            perPage: 1,
            page: null,
            sortBy: null,
            filter: null
        };
    };
    Standing.parse = function (data) {
        return new Standing(data.id, data.placement, Entrant_1.Entrant.parse(data.entrant));
    };
    Standing.prototype.getId = function () {
        return this.id;
    };
    Standing.prototype.getPlacement = function () {
        return this.placement;
    };
    Standing.prototype.getEntrant = function () {
        return this.entrant;
    };
    Standing.prototype.getGamerTag = function () {
        if (this.entrant)
            return this.entrant.getAttendee().getGamerTag();
        return null;
    };
    return Standing;
}());
exports.Standing = Standing;
