// smashgg.js/src/v4/js/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Initializer_1 = require("./lib/util/Initializer");
exports.initialize = Initializer_1.default;
var GGSet_1 = require("./lib/models/GGSet");
exports.GGSet = GGSet_1.GGSet;
var Event_1 = require("./lib/models/Event");
exports.Event = Event_1.Event;
var Phase_1 = require("./lib/models/Phase");
exports.Phase = Phase_1.Phase;
var Entrant_1 = require("./lib/models/Entrant");
exports.Entrant = Entrant_1.Entrant;
var Attendee_1 = require("./lib/models/Attendee");
exports.Attendee = Attendee_1.Attendee;
var Organizer_1 = require("./lib/models/Organizer");
exports.Organizer = Organizer_1.Organizer;
var Venue_1 = require("./lib/models/Venue");
exports.Venue = Venue_1.Venue;
var Stream_1 = require("./lib/models/Stream");
exports.Stream = Stream_1.Stream;
var StreamQueue_1 = require("./lib/models/StreamQueue");
exports.StreamQueue = StreamQueue_1.StreamQueue;
var Tournament_1 = require("./lib/models/Tournament");
exports.Tournament = Tournament_1.Tournament;
var PhaseGroup_1 = require("./lib/models/PhaseGroup");
exports.PhaseGroup = PhaseGroup_1.PhaseGroup;
var VideoGame_1 = require("./lib/models/VideoGame");
exports.VideoGame = VideoGame_1.VideoGame;
var Character_1 = require("./lib/models/Character");
exports.Character = Character_1.Character;
var Logger_1 = require("./lib/util/Logger");
exports.Log = Logger_1.default;
exports.setLogLevel = Logger_1.setLogLevel;
exports.addLog = Logger_1.addLog;
exports.disableLog = Logger_1.disableLog;
exports.enableLog = Logger_1.enableLog;
